ext/drx/scatter_gather_aarch64.c

```diff
@@ -595,12 +595,36 @@ expand_contiguous(void *drcontext, instrlist_t *bb, instr_t *sg_instr,
      * BSD licensed drx to have a dependency on the LGPL licensed drutil.
      */
 
-    DR_ASSERT(sg_info->index_reg != DR_REG_NULL);
+    reg_id_t new_base = scratch_gpr0;
 
-    /* add      scratch_gpr0, base_reg, index_reg, extend #extend_amount */
-    EMIT(add_extend, opnd_create_reg(scratch_gpr0), opnd_create_reg(sg_info->base_reg),
-         opnd_create_reg(sg_info->index_reg), OPND_CREATE_INT(sg_info->extend),
-         OPND_CREATE_INT(sg_info->extend_amount));
+    if (sg_info->index_reg == DR_REG_NULL) {
+        /* scalar+immediate */
+        DR_ASSERT(!sg_info->scaled);
+        DR_ASSERT(sg_info->extend_amount == 0);
+
+        if (sg_info->disp == 0) {
+            /* The displacement is 0 so the original base register already contains the
+             * base of the contiguous memory region.
+             */
+            new_base = sg_info->base_reg;
+        } else if (sg_info->disp > 0) {
+            /* add      new_base, base_reg, #disp */
+            EMIT(add, opnd_create_reg(new_base), opnd_create_reg(sg_info->base_reg),
+                 OPND_CREATE_INT(sg_info->disp));
+        } else {
+            /* sub      new_base, base_reg, #-disp */
```

Comment:
It's subtracting the negative displacement; double negatives turning it into addition? Is there a correctness issue?

Reply:
> It's subtracting the negative displacement; double negatives turning it into addition?

In a way.

The `disp` value is a signed displacement that is added to the base, however the AArch64 `add` and `sub` instructions use an unsigned immediate value so we can't use `add` to add a negative `disp`.

For example, if `disp` was `-4`, what we want to do is:
```
add   x0, x1, #-4 ; Invalid instruction
```
but we can't so we turn it into
```
sub   x0, x1, #4 ; Subtracting 4 is the same as adding -4
```